<h1>Soal 1: Pemahaman</h1>

1. Sebutkan apa saja yang termasuk preprocessing data!
2. Jelaskan beberapa cara imputing missing value!
3. Kapan kita perlu melakukan feature centering dan scaling?
4. Bagaimana Data Science Workflow?

---

<h1>Soal 2: Pengaplikasian</h1>

Selamat, sampai tahap ini kalian telah belajar banyak tentang data science, dari mulai python, data manipulasi, visualisasi, dan pembuatan model. Sekarang saatnya untuk mengaplikasikan semuanya.

Download dan gunakan data [titanic.csv](https://drive.google.com/uc?export=download&id=15-XQkmqj2UlFQH7rASJH1oxWST6o9mbm) sebagai data untuk pembuatan model ML. Pahami betul data ini dengan melakukan EDA (Explolatory Data Analaysis), Visualisasi, Data Analysis, Preprocessing Data, dan Modeling.

<b>(Optional)</b> Download dan gunakan data [titanic_test.csv](https://drive.google.com/uc?export=download&id=15-XQkmqj2UlFQH7rASJH1oxWST6o9mbm) untuk mengetest model kalian dengan melakukan prediksi terhadap data tersebut. Submit hasil prediksinya ke kaggle dan lihat scorenya. https://www.kaggle.com/c/titanic/submit

![alt text](https://drive.google.com/uc?id=1SyflO2YeuCdYhxwPF6CAtMeaNN1qTaFx)

In [1]:
# Read data
import pandas as pd

df = pd.read_csv('titanic.csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [2]:
df[['PassengerId','Survived','Pclass','Name','Sex','Age','SibSp','Parch','Ticket','Fare','Cabin','Embarked']].isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [3]:
#Menghapus data unique
df.drop(['Name','Ticket'], inplace = True, axis = 1)

In [4]:
#tampilan data yang kita miliki sekarang
df.head()

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin  \
0            1         0       3    male  22.0      1      0   7.2500   NaN   
1            2         1       1  female  38.0      1      0  71.2833   C85   
2            3         1       3  female  26.0      0      0   7.9250   NaN   
3            4         1       1  female  35.0      1      0  53.1000  C123   
4            5         0       3    male  35.0      0      0   8.0500   NaN   

  Embarked  
0        S  
1        C  
2        S  
3        S  
4        S

In [5]:
#membersihkan data tidak clean pada sex dan embarked
#mencari nilai unique pada masing-masing kolom
print(df['Sex'].unique())
print(df['Embarked'].unique())

['male' 'female']
['S' 'C' 'Q' nan]


In [6]:
obj_sex = {'male' : 0, 'female' : 1}

obj_embarked = {'S': 0, 'C': 1, 'Q': 2}

df['Sex'].replace(obj_sex, inplace = True)
df['Embarked'].replace(obj_embarked, inplace = True)

In [7]:
#tampilan data yang kita miliki sekarang
df.head()

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare Cabin  \
0            1         0       3    0  22.0      1      0   7.2500   NaN   
1            2         1       1    1  38.0      1      0  71.2833   C85   
2            3         1       3    1  26.0      0      0   7.9250   NaN   
3            4         1       1    1  35.0      1      0  53.1000  C123   
4            5         0       3    0  35.0      0      0   8.0500   NaN   

   Embarked  
0       0.0  
1       1.0  
2       0.0  
3       0.0  
4       0.0

In [8]:
#membersihkan data tidak clean pada age, cabin, dan embarked
#mengetahui shape data
df.shape

(891, 10)

Terdapat 891 baris, diketahui dari data tadi bahwa missing values dari kolom Age = 177, Cabin = 687, Embarked = 2, Sehingga tidak mungkin melakukan drop pada kolom umur karena jumlahnya terlalu banyak dan menurut saya kolom dan baris tentang umur penting dalam melakukan pengolahan data. Kolom cabin juga berpengaruh terhadap pengolahan data. Sehingga saya memutuskan menggunakan KNN karena kemungkinan terdapat korelasi antara kolom tersebut

In [9]:
#mengubah data NaN pada cabin menjadi '0'
import numpy as np

df['Cabin'].replace(np.nan, '0', inplace = True)

#mencari nilai unique pada kolom cabin
print(df['Cabin'].unique())

['0' 'C85' 'C123' 'E46' 'G6' 'C103' 'D56' 'A6' 'C23 C25 C27' 'B78' 'D33'
 'B30' 'C52' 'B28' 'C83' 'F33' 'F G73' 'E31' 'A5' 'D10 D12' 'D26' 'C110'
 'B58 B60' 'E101' 'F E69' 'D47' 'B86' 'F2' 'C2' 'E33' 'B19' 'A7' 'C49'
 'F4' 'A32' 'B4' 'B80' 'A31' 'D36' 'D15' 'C93' 'C78' 'D35' 'C87' 'B77'
 'E67' 'B94' 'C125' 'C99' 'C118' 'D7' 'A19' 'B49' 'D' 'C22 C26' 'C106'
 'C65' 'E36' 'C54' 'B57 B59 B63 B66' 'C7' 'E34' 'C32' 'B18' 'C124' 'C91'
 'E40' 'T' 'C128' 'D37' 'B35' 'E50' 'C82' 'B96 B98' 'E10' 'E44' 'A34'
 'C104' 'C111' 'C92' 'E38' 'D21' 'E12' 'E63' 'A14' 'B37' 'C30' 'D20' 'B79'
 'E25' 'D46' 'B73' 'C95' 'B38' 'B39' 'B22' 'C86' 'C70' 'A16' 'C101' 'C68'
 'A10' 'E68' 'B41' 'A20' 'D19' 'D50' 'D9' 'A23' 'B50' 'A26' 'D48' 'E58'
 'C126' 'B71' 'B51 B53 B55' 'D49' 'B5' 'B20' 'F G63' 'C62 C64' 'E24' 'C90'
 'C45' 'E8' 'B101' 'D45' 'C46' 'D30' 'E121' 'D11' 'E77' 'F38' 'B3' 'D6'
 'B82 B84' 'D17' 'A36' 'B102' 'B69' 'E49' 'C47' 'D28' 'E17' 'A24' 'C50'
 'B42' 'C148']


In [10]:
#mengubah data kolom cabing dari string menjadi float
k = df['Cabin'].unique()
v = np.arange(0, len(k))

obj_cabin = dict(zip(k,v))

df['Cabin'] = df['Cabin'].replace(obj_cabin)
df['Cabin'] = df['Cabin'].replace(0, np.nan)

In [11]:
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors = 3)
miss_value = ['Age','Cabin','Embarked']
df[miss_value] = imputer.fit_transform(df[miss_value])

df

PassengerId  Survived  Pclass  Sex        Age  SibSp  Parch     Fare  \
0              1         0       3    0  22.000000      1      0   7.2500   
1              2         1       1    1  38.000000      1      0  71.2833   
2              3         1       3    1  26.000000      0      0   7.9250   
3              4         1       1    1  35.000000      1      0  53.1000   
4              5         0       3    0  35.000000      0      0   8.0500   
..           ...       ...     ...  ...        ...    ...    ...      ...   
886          887         0       2    0  27.000000      0      0  13.0000   
887          888         1       1    1  19.000000      0      0  30.0000   
888          889         0       3    1  27.666667      1      2  23.4500   
889          890         1       1    0  26.000000      0      0  30.0000   
890          891         0       3    0  32.000000      0      0   7.7500   

          Cabin  Embarked  
0     23.000000       0.0  
1      1.000000       1.0  
2     25.000000       0.0  
3      2.000000       0.0  
4     14.333333       0.0  
..          ...       ...  
886   25.000000       0.0  
887  146.000000       0.0  
888    3.000000       0.0  
889  147.000000       1.0  
890   64.666667       2.0  

[891 rows x 10 columns]

In [12]:
# mengetahui jumlah orang yang selamat
df['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

In [13]:
#menentukan model dari data, disini saya menggunakan KNeighborsClassifier 
from sklearn.neighbors import KNeighborsClassifier

X = df.drop('Survived', axis = 1)
y = df['Survived']

In [14]:
#preprocessing data
from sklearn.preprocessing import StandardScaler

stds = StandardScaler()
datastd = stds.fit_transform(X)
X = pd.DataFrame(datastd, columns = X.columns)

X

PassengerId    Pclass       Sex       Age     SibSp     Parch      Fare  \
0      -1.730108  0.827377 -0.737695 -0.481016  0.432793 -0.473674 -0.502445   
1      -1.726220 -1.566107  1.355574  0.694564  0.432793 -0.473674  0.786845   
2      -1.722332  0.827377  1.355574 -0.187121 -0.474545 -0.473674 -0.488854   
3      -1.718444 -1.566107  1.355574  0.474143  0.432793 -0.473674  0.420730   
4      -1.714556  0.827377 -0.737695  0.474143 -0.474545 -0.473674 -0.486337   
..           ...       ...       ...       ...       ...       ...       ...   
886     1.714556 -0.369365 -0.737695 -0.113647 -0.474545 -0.473674 -0.386671   
887     1.718444 -1.566107  1.355574 -0.701437 -0.474545 -0.473674 -0.044381   
888     1.722332  0.827377  1.355574 -0.064664  0.432793  2.008933 -0.176263   
889     1.726220 -1.566107 -0.737695 -0.187121 -0.474545 -0.473674 -0.044381   
890     1.730108  0.827377 -0.737695  0.253722 -0.474545 -0.473674 -0.492378   

        Cabin  Embarked  
0   -0.369537 -0.568837  
1   -1.095404  1.005181  
2   -0.303549 -0.568837  
3   -1.062410 -0.568837  
4   -0.655484 -0.568837  
..        ...       ...  
886 -0.303549 -0.568837  
887  3.688717 -0.568837  
888 -1.029416 -0.568837  
889  3.721711  1.005181  
890  1.005210  2.579199  

[891 rows x 9 columns]

In [15]:
#menentukan hyperparameter
from sklearn.model_selection import GridSearchCV

model = KNeighborsClassifier()
param_grid = {
              'n_neighbors' : np.arange(1,50), 
              'weights': ['distance','uniform'],
              'algorithm' : ['auto','ball_tree','kd_tree','brute']
             }

gscv = GridSearchCV(model, param_grid, scoring = 'accuracy', cv = 5)
gscv.fit(X,y)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],
                         'n_neighbors': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]),
                         'weights': ['distance', 'uniform']},
             scoring='accuracy')

In [16]:
gscv.best_params_

{'algorithm': 'auto', 'n_neighbors': 10, 'weights': 'uniform'}

In [17]:
gscv.best_score_

0.8170547988199107

In [18]:
model_knn = KNeighborsClassifier(n_neighbors = 10, algorithm = 'auto', weights = 'uniform')

Setelah menentukan model, kita akan melakukan prediksi terhadap data test

In [19]:
df_test = pd.read_csv('titanic_test.csv')
df_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [20]:
df_test[['PassengerId','Pclass','Name','Sex','Age','SibSp','Parch','Ticket','Fare','Cabin','Embarked']].isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [21]:
#melakukan perulangan codingnya sesuai dengan data training sebelumnya
df_test.drop(['Name','Ticket'], inplace = True, axis = 1)

In [22]:
df_test['Sex'].replace(obj_sex, inplace = True)
df_test['Embarked'].replace(obj_embarked, inplace = True)

In [23]:
df_test['Cabin'].replace(np.nan, '0', inplace = True)

k = df_test['Cabin'].unique()
v = np.arange(0, len(k))

obj_cabin = dict(zip(k,v))

df_test['Cabin'] = df_test['Cabin'].replace(obj_cabin)
df_test['Cabin'] = df_test['Cabin'].replace(0, np.nan)

df_test

PassengerId  Pclass  Sex   Age  SibSp  Parch      Fare  Cabin  Embarked
0            892       3    0  34.5      0      0    7.8292    NaN         2
1            893       3    1  47.0      1      0    7.0000    NaN         0
2            894       2    0  62.0      0      0    9.6875    NaN         2
3            895       3    0  27.0      0      0    8.6625    NaN         0
4            896       3    1  22.0      1      1   12.2875    NaN         0
..           ...     ...  ...   ...    ...    ...       ...    ...       ...
413         1305       3    0   NaN      0      0    8.0500    NaN         0
414         1306       1    1  39.0      0      0  108.9000   76.0         1
415         1307       3    0  38.5      0      0    7.2500    NaN         0
416         1308       3    0   NaN      0      0    8.0500    NaN         0
417         1309       3    0   NaN      1      1   22.3583    NaN         1

[418 rows x 9 columns]

In [24]:
miss_value2 = ['Age','Fare','Fare','Cabin']
df_test[miss_value2] = imputer.fit_transform(df_test[miss_value2])

df_test

PassengerId  Pclass  Sex        Age  SibSp  Parch      Fare      Cabin  \
0            892       3    0  34.500000      0      0    7.8292  41.000000   
1            893       3    1  47.000000      1      0    7.0000  57.333333   
2            894       2    0  62.000000      0      0    9.6875  57.333333   
3            895       3    0  27.000000      0      0    8.6625  41.000000   
4            896       3    1  22.000000      1      1   12.2875  48.000000   
..           ...     ...  ...        ...    ...    ...       ...        ...   
413         1305       3    0  25.000000      0      0    8.0500  41.000000   
414         1306       1    1  39.000000      0      0  108.9000  76.000000   
415         1307       3    0  38.500000      0      0    7.2500  54.333333   
416         1308       3    0  25.000000      0      0    8.0500  41.000000   
417         1309       3    0  28.666667      1      1   22.3583  48.666667   

     Embarked  
0           2  
1           0  
2           2  
3           0  
4           0  
..        ...  
413         0  
414         1  
415         0  
416         0  
417         1  

[418 rows x 9 columns]

In [25]:
datastd = stds.fit_transform(df_test)
X_test = pd.DataFrame(datastd, columns = df_test.columns)

X_test

PassengerId    Pclass       Sex       Age     SibSp     Parch      Fare  \
0      -1.727912  0.873482 -0.755929  0.390251 -0.499470 -0.400248 -0.500288   
1      -1.719625  0.873482  1.322876  1.346359  0.616992 -0.400248 -0.515002   
2      -1.711337 -0.315819 -0.755929  2.493689 -0.499470 -0.400248 -0.467312   
3      -1.703050  0.873482 -0.755929 -0.183414 -0.499470 -0.400248 -0.485501   
4      -1.694763  0.873482  1.322876 -0.565857  0.616992  0.619896 -0.421174   
..           ...       ...       ...       ...       ...       ...       ...   
413     1.694763  0.873482 -0.755929 -0.336391 -0.499470 -0.400248 -0.496370   
414     1.703050 -1.505120  1.322876  0.734450 -0.499470 -0.400248  1.293253   
415     1.711337  0.873482 -0.755929  0.696205 -0.499470 -0.400248 -0.510566   
416     1.719625  0.873482 -0.755929 -0.336391 -0.499470 -0.400248 -0.496370   
417     1.727912  0.873482 -0.755929 -0.055933  0.616992  0.619896 -0.242463   

        Cabin  Embarked  
0    0.116744  2.243165  
1    1.272447 -0.677841  
2    1.272447  2.243165  
3    0.116744 -0.677841  
4    0.612045 -0.677841  
..        ...       ...  
413  0.116744 -0.677841  
414  2.593251  0.782662  
415  1.060175 -0.677841  
416  0.116744 -0.677841  
417  0.659217  0.782662  

[418 rows x 9 columns]

In [26]:
# memulai prediksi
y_pred = gscv.predict(X_test)
df_test['Survived'] = y_pred

df_test

PassengerId  Pclass  Sex        Age  SibSp  Parch      Fare      Cabin  \
0            892       3    0  34.500000      0      0    7.8292  41.000000   
1            893       3    1  47.000000      1      0    7.0000  57.333333   
2            894       2    0  62.000000      0      0    9.6875  57.333333   
3            895       3    0  27.000000      0      0    8.6625  41.000000   
4            896       3    1  22.000000      1      1   12.2875  48.000000   
..           ...     ...  ...        ...    ...    ...       ...        ...   
413         1305       3    0  25.000000      0      0    8.0500  41.000000   
414         1306       1    1  39.000000      0      0  108.9000  76.000000   
415         1307       3    0  38.500000      0      0    7.2500  54.333333   
416         1308       3    0  25.000000      0      0    8.0500  41.000000   
417         1309       3    0  28.666667      1      1   22.3583  48.666667   

     Embarked  Survived  
0           2         0  
1           0         1  
2           2         0  
3           0         0  
4           0         0  
..        ...       ...  
413         0         0  
414         1         1  
415         0         0  
416         0         0  
417         1         0  

[418 rows x 10 columns]

In [27]:
result = df_test[['PassengerId','Survived']]
result

PassengerId  Survived
0            892         0
1            893         1
2            894         0
3            895         0
4            896         0
..           ...       ...
413         1305         0
414         1306         1
415         1307         0
416         1308         0
417         1309         0

[418 rows x 2 columns]

In [28]:
result.to_csv('titanic_test_final', index = False)